pkg/apis/core/v1alpha1/workflow_methods.go

```diff
@@ -30,3 +30,83 @@ func (w *Workflow) GetStartTime() *metav1.Time {
 	}
 	return nil
 }
+
+type taskInfo struct {
+	CurrentWorker        string
+	CurrentTask          string
+	CurrentTaskIndex     int
+	CurrentAction        string
+	CurrentActionIndex   int
+	CurrentActionState   string
+	TotalNumberOfActions int
+}
+
+// helper function for task info.
+func (w *Workflow) getTaskActionInfo() taskInfo {
+	var (
+		found           bool
+		taskIndex       = -1
+		actionIndex     int
+		actionTaskIndex int
+		actionCount     int
+	)
+	for ti, task := range w.Status.Tasks {
+		for ai, action := range task.Actions {
+			actionCount++
+			// Find either the first action in a non-terminal state or the first action in an unsuccessful terminal state
+			if (action.Status == "STATE_PENDING" || action.Status == "STATE_RUNNING" ||
+				action.Status == "STATE_FAILED" || action.Status == "STATE_TIMEOUT") && !found {
+				taskIndex = ti
+				actionTaskIndex = ai
+				found = true
+			}
+			if !found {
+				actionIndex++
+			}
+		}
+	}
```

Comment:
I was having some difficulty reading through the original logic, so I took a swag at trying to make it a bit more readable, not sure if I succeeded, though.

```suggestion
	taskIndex, actionIndex, actionTaskIndex, actionCount := func (tasks []Tasks) (int, int, int, int) {
		var (
			found           bool
			taskIndex       = -1
			actionIndex     int
			actionTaskIndex int
			actionCount     int
		)
		
		for ti, task := range w.Status.Tasks {
			actionCount += len(task.Actions)
			
			if found {
				continue
			}
			
			for ai, action := range task.Actions {
				switch action.Status {
				case "STATE_SUCCESS":
					actionIndex++
					continue	
				case "STATE_PENDING", "STATE_RUNNING", "STATE_FAILED", "STATE_TIMEOUT":
					taskIndex = ti
					actionTaskIndex = ai
					found = true
					break
				}
			}
		}
		
		return taskIndex, actionIndex, actionTaskIndex, actionCount
	}()
```

Reply:
Good suggestion, and the tests pass with this. thank you!